### PR TITLE
[CALCITE-6742] StandardConvertletTable.convertCall loses casts from ROW comparisons

### PR DESCRIPTION
**Pull request by mihaibudiu**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java [hunk 1]
```diff
@@ -563,9 +563,15 @@ private Expression getConvertExpression(
     case TINYINT:
     case SMALLINT: {
       if (SqlTypeName.NUMERIC_TYPES.contains(sourceType.getSqlTypeName())) {
+        Type javaClass = typeFactory.getJavaClass(targetType);
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:566]: This change is only indirectly related to this PR: without this change, the code generated for ROW casts crashes at runtime.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java:566]: Why does it crash? It feels a bit strange

**Comment by mihaibudiu** [comment]
This PR fixes a real bug and should not be too hard to review.

**Review by caicancai — APPROVED** [review]

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 2]
```diff
@@ -7654,6 +7654,28 @@ from orders]]>
...
+]]>
+    </Resource>
+  </TestCase>
+  <TestCase name="testStructCast1">
+    <Resource name="sql">
+      <![CDATA[select CAST(CAST(ROW('x', 1) AS ROW(l INTEGER, r DOUBLE)) AS ROW(l BIGINT, r INTEGER)) = ROW(RAND(), RAND())]]>
+    </Resource>
+    <Resource name="plan">
+      <![CDATA[
+LogicalProject(EXPR$0=[AND(=(CAST(CAST('x'):INTEGER NOT NULL):DOUBLE NOT NULL, RAND()), =(1.0E0, RAND()))])
```
Comment (ILuffZhe) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:7677]: `1.0E0` comes from `cast 1 as DOUBLE`?

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 3]
```diff
@@ -7654,6 +7654,28 @@ from orders]]>
 LogicalDelta
   LogicalProject(ROWTIME=[$0], PRODUCTID=[$1], ORDERID=[$2], C=[COUNT() OVER (PARTITION BY $1 ORDER BY $0 RANGE 1000:INTERVAL SECOND PRECEDING)])
     LogicalTableScan(table=[[CATALOG, SALES, ORDERS]])
+]]>
+    </Resource>
+  </TestCase>
+  <TestCase name="testStructCast">
+    <Resource name="sql">
+      <![CDATA[select ROW(1, 'x') = ROW('x', 1)]]>
```
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:7662]: This becomes `CAST('x' AS INTEGER) = 1 AND 1 = CAST('x' AS INTEGER)`, which is then simplified

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml (hunk 2 above)
Comment (mihaibudiu) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:7677]: yes, there are many places where Calcite simplifies expressions, it's hard to write a test exhibits just the change I made.

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mihaibudiu) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:7662]: I will change the test the way you suggested

[the remark this answers is not in the document]
Comment (ILuffZhe) [reply core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml]: Got it. Thanks for the illustration.

**Pull request merged**